### DISCUSSION TITLE
Support ssh keys for artifacts updating

### DISCUSSION THREAD

**Discussion by mettke** [renovatebot/renovate#24928] [Suggest an Idea]
### What would you like Renovate to be able to do?

We want to be able to include private dependencies via ssh. Using ssh is a lot more convenient then using https, as developers don't have to add their github/gitlab credentials to their machine but instead can use their ssh keys.

SSH currently does not work as there is no way to add a private key to renovate with the permissions to check out the repositories. If it were possible to add a private key, package manager would be able to pick it up and work out of the box.

### If you have any ideas on how this should be implemented, please tell us here.

Renovate already offers options to add encrypted credentials to the renovate config. We could extend it by also allowing the user to add an encrypted ssh private key (potentially even multiple but that could be a bit more difficult). 

On renovate side, we would need to:
1. Decrypt the private key
2. Add the key to `~/.ssh/id_rsa`

From there on, package manager should be able to pick it up and us it. We, however, have to make sure that the system does not hang on asking for a password, if the user submits a password encrypted ssh key.

This solution would work with github and gitlab, as they require deploy keys to be created locally with only the public key being submitted. In other words, the user could generate one key and add it to multiple repositories. 

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by mettke** [comment]
I'm happy to take a shot at this myself, if someone is available for a bit of mentoring.

**Comment by viceice** [comment]
Writing to `~/.ssh/id_rsa` is a bad idea, it can overwrite existing keys when running renovate locally. Instead we should use ssh-agent amd forward the socket to the child container

- https://github.com/renovatebot/renovate/discussions/11575#discussioncomment-1323672
- https://github.com/renovatebot/renovate/discussions/11745#discussioncomment-1325321

**Comment by rarkins** [comment]
Could also be isolated to sidecar containers only for the initial version?

**Comment by viceice** [comment]
Sure, I think best is when renovate starts the ssh-agent with a fixed socket path ( on cache volume, which is mounted to docker), so it's usable by all renovate child commands.
After each repo the agent should be cleared / restarted.

**Comment by viceice** [comment]
pure js ssh-agents

- https://www.npmjs.com/package/ssh-agent
- https://www.npmjs.com/package/ssh-agent-js

**Comment by HonkingGoose** [comment]
## Background

We closed @mettke's old PR because it needed more work, and was getting stale and conflicted. Revant quotes:

> Needs documentation, potentially both internal (for future maintainers) as well as external.
> 
> I'm curious why the agent needs to be passed to the renovateRepository() command. Seems like it's going to be passed every time, so why can't it just live in its own module and be accessed directly by logic within renovateRepository()?

> Closed due to inactivity and conflicts. Welcome to reopen/recreate in future

## Link to old PR

- #14963

## Questions for @mettke 

Do you still want to work on this @mettke, or do you want somebody else to try? It looks like want somebody to mentor you through this PR?

**Comment by mettke** [comment]
@HonkingGoose I tried implementing it but for me it was very difficult to work through the tests to get them up and running (with my changes). I had a lot of flakyness which probably was introduced due to me spawning an external command.

I do not think that I will pick this up again, even though I still think that this would be a nice addition. In general my code does work already (tried it manually), only the tests need more work.
